python/lsst/ts/m2com/controller_cell.py

```diff
@@ -0,0 +1,475 @@
+        process_event : `func` or `coroutine`
+            Function to process the event. It must has a keyward argument of
+            "message" to receive the message as a dictionary.
+        *args : `args`
+            Arguments needed in "process_event" function call.
+        **kwargs : `dict`, optional
+            Additional keyword arguments in "process_event" function
+            call.
+        """
+        self._task_event_loop = asyncio.create_task(
+            self._event_loop(process_event, *args, **kwargs)
+        )
+
+    async def _event_loop(self, process_event, *args, **kwargs):
+        """Update and output event information from component.
+
+        Parameters
+        ----------
+        process_event : `func` or `coroutine`
+            Function to process the event. It must has a keyward argument of
+            "message" to receive the message as a dictionary.
+        *args : `args`
+            Arguments needed in "process_event" function call.
+        **kwargs : `dict`, optional
+            Additional keyword arguments in "process_event" function
+            call.
+        """
+
+        self.log.debug("Begin to run the event loop from component.")
+
+        is_coroutine_function = is_coroutine(process_event)
+
+        while self.run_loops:
+
+            try:
+                message = (
+                    self.queue_event.get_nowait()
+                    if not self.queue_event.empty()
+                    else await self.queue_event.get()
+                )
+
+            # When there is no instance in the self.queue_event, the get() will
+            # be used to wait for the new coming event. It is a blocking call
+            # and will get the asyncio.CancelledError if we cancel the
+            # self._task_event_loop.
+            except asyncio.CancelledError:
+                message = ""
+
+            # Process the event
+            if is_coroutine_function:
+                await process_event(*args, message=message, **kwargs)
+            else:
+                process_event(*args, message=message, **kwargs)
+
+            await asyncio.sleep(self.timeout)
+
+        self.log.debug("Component's event loop exited.")
+
+    def run_task_telemetry_loop(self, process_telemetry, *args, period=2, **kwargs):
+        """Run the task of telemetry loop.
+
+        Parameters
+        ----------
+        process_telemetry : `func` or `coroutine`
+            Function to process the telemetry. It must has a keyward argument
+            of "message" to receive the message as a dictionary.
+        *args : `args`
+            Arguments needed in "process_telemetry" function call.
+        period : `float` or `int`, optional
+            Period to check the telemetry rate in second. (the default is 2)
+        **kwargs : `dict`, optional
+            Additional keyword arguments in "process_telemetry" function
+            call.
+        """
+        self._task_telemetry_loop = asyncio.create_task(
```

Comment:
This is starting the task without checking that the task is already running. Consider adding some check here that this is safe to do. I am not really sure what is going to happen if you call this method twice, but I don't think it is going to be pretty. 

Also, I think you may have to convert this into a coroutine, so you can actually manage it here. Unless you want to fail if the task is already running. Something like this would be the simplest thing to do:

```
if self._task_telemetry_loop is not None:
    raise RuntimeError("Telemetry loop task is already running or was already started.")
```

But you may want to create ways for the "user" to clean it up as well.

Reply:
In the `ControllerCell.__init__()`, the task attributes are initiated to be the `Future()` instance already by `lsst.ts.utils.make_done_future()`. There is the `ControllerCell.stop_loops()` already to cancel/close all the running tasks. Thanks!